src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -27,8 +25,50 @@ public ChangedText(SourceText oldText, SourceText newText, ImmutableArray<TextCh
             Debug.Assert(!changeRanges.IsDefault);
 
             _newText = newText;
-            _weakOldText = new WeakReference<SourceText>(oldText);
-            _changes = changeRanges;
+            _info = new ChangeInfo(changeRanges, new WeakReference<SourceText>(oldText), (oldText as ChangedText)?._info);
+        }
+
+        private class ChangeInfo
+        {
+            public ImmutableArray<TextChangeRange> ChangeRanges { get; }
+
+            // store old text weakly so we don't form unwanted chains of old texts (especially chains of ChangedTexts)
+            // used to identify the changes in GetChangeRanges.
+            public WeakReference<SourceText> WeakOldText { get; }
+
+            public ChangeInfo Previous { get; private set; }
+
+            public ChangeInfo(ImmutableArray<TextChangeRange> changeRanges, WeakReference<SourceText> weakOldText, ChangeInfo previous)
+            {
+                this.ChangeRanges = changeRanges;
+                this.WeakOldText = weakOldText;
+                this.Previous = previous;
+                Clean();
+            }
+
+            // clean up 
+            private void Clean()
+            {
+                // look for last info in the chain that still has reference to old text
+                ChangeInfo lastInfo = this;
+                for (var info = this; info != null; info = info.Previous)
```

Comment:
Yes, a ChangedText caught in gen2 hell may keep a chain of ChangeInfo's alive and the cost of checking would increase over time if the changes are on going.   However, you'll never actually do the merges that far back because you don't have a reference to any of those source texts..  You only do the merges on the changes between the two SourceTexts that you references to.